### PR TITLE
Issue#316 : ParserFileHelperFactory Class Refactored

### LINKED ISSUE #316: Refactor getCorrespondingParserFileHelper to reduce cognitive complexity
https://sonarcloud.io/project/issues?id=camel-lsp-server&issues=AWxRVsbydg2jiL0y_QTx&open=AWxRVsbydg2jiL0y_QTx

Screenshot from 2019-10-04 15-53-45

### PR DESCRIPTION
**Pull request by kapilgahlot1998**
**getCorrespondingParserFileHelper** method of **ParserFileHelperFactory** class refactored for cognitive complexity.

Multiple if-else condition block of code changed with **Chain Of Responsibility** Pattern

Fixed #316

### CODE REVIEW

File: src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/XMLFileParser.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    @Override
+    public ParserFileHelper getMatchedFileParser(TextDocumentItem textDocumentItem, int line, ParserFileHelperFactory parserFileHelperFactory){
+        ParserXMLFileHelper xmlParser = new ParserXMLFileHelper();
+        String uri = textDocumentItem.getUri();
+        if (uri.endsWith(".xml") && xmlParser.getCorrespondingCamelNodeForCompletion(textDocumentItem, line) != null)
+            return xmlParser;
+        else if(nextFileParser != null)
+            return nextFileParser.getMatchedFileParser(textDocumentItem, line, parserFileHelperFactory);
```
Comment (apupier) [inline src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/XMLFileParser.java:39]: Please add brackets. I've hit so many dummy bugs in my experience due to missing brackets that I really prefer to have them always.

**Review by apupier — CHANGES_REQUESTED** [review]
this is a good improvement, thank you.
Only small style changes required.

Please update the PR to use tabs instead of space for indentation. it allows to stay coherent with other files in this project and i tis a good practice to be more inclusive for users that requires specific alignment or screen readers.

File: src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/XMLFileParser.java (hunk 1 above)
Comment (kapilgahlot1998) [reply src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/XMLFileParser.java:39]: OK sure will update this request. Thanks

Comment (kapilgahlot1998) [reply src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/XMLFileParser.java]: @apupier  Code refactoring for the adding braces is done. 

Also I am using tab indentation only. I checked my idea setting its 4 spaces indentation only. Let me know in the code if you find any place where the indentation is not proper. I will update.

File: src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/CamelKGroovyDSLFileParser.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            CamelKGroovyDSLParser camelKGroovyDSLParser = new CamelKGroovyDSLParser();
+            if (camelKGroovyDSLParser.getCorrespondingMethodName(textDocumentItem, line) != null) {
+                return camelKGroovyDSLParser;
+            }
+        } else if(nextFileParser != null) {
+            return nextFileParser.getMatchedFileParser(textDocumentItem, line, parserFileHelperFactory);
+        }
+
+        return null;
+    }
```
Comment (apupier) [inline src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/CamelKGroovyDSLFileParser.java:45]: spaces used here

File: src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/CamelKJSDSLFileParser.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            CamelKJSDSLParser camelKJSDSLParser = new CamelKJSDSLParser();
+            if (camelKJSDSLParser.getCorrespondingMethodName(textDocumentItem, line) != null) {
+                return camelKJSDSLParser;
+            }
+        } else if(nextFileParser != null) {
+            return nextFileParser.getMatchedFileParser(textDocumentItem, line, parserFileHelperFactory);
+        }
+
+        return null;
+    }
```
Comment (apupier) [inline src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/CamelKJSDSLFileParser.java:45]: spaces used here

File: src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/CamelKKotlinDSLFileParser.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            CamelKKotlinDSLParser camelKKotlinDSLParser = new CamelKKotlinDSLParser();
+            if (camelKKotlinDSLParser.getCorrespondingMethodName(textDocumentItem, line) != null) {
+                return camelKKotlinDSLParser;
+            }
+        } else if(nextFileParser != null) {
+            return nextFileParser.getMatchedFileParser(textDocumentItem, line, parserFileHelperFactory);
+        }
+
+        return null;
+    }
```
Comment (apupier) [inline src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/CamelKKotlinDSLFileParser.java:45]: spaces used here

File: src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/XMLFileParser.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        ParserXMLFileHelper xmlParser = new ParserXMLFileHelper();
+        String uri = textDocumentItem.getUri();
+        if (uri.endsWith(".xml") && xmlParser.getCorrespondingCamelNodeForCompletion(textDocumentItem, line) != null) {
+            return xmlParser;
+        } else if(nextFileParser != null) {
+            return nextFileParser.getMatchedFileParser(textDocumentItem, line, parserFileHelperFactory);
+        }
+
+        return null;
+    }
```
Comment (apupier) [inline src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/XMLFileParser.java:43]: spaces used here

Comment (kapilgahlot1998) [reply src/main/java/com/github/cameltooling/lsp/internal/parser/fileparserhelper/XMLFileParser.java]: @apupier  I have updated the MR with tab indentation.

**Review by apupier — APPROVED** [review]
Thank you for your contribution

**Comment by kapilgahlot1998** [comment]
@apupier Thanks a lot for guiding me. Will check other issues If I can resolve.

**Pull request merged**